### PR TITLE
fix l1 fee rate selection and error handling in sequencer

### PR DESCRIPTION
**Pull request by eyusufatik**
# Description
Sequencer now checks l1 fee rate is in the allowable range by SoftConfirmationRuleEnforcer.

Also we realized the produce_l2block does not return error on begin_soft_confirmation_hook error, which causes problems

### CODE REVIEW

File: crates/sequencer/src/sequencer.rs [hunk 1]
```diff
@@ -464,12 +474,22 @@ where
...
-        let l1_fee_rate = self
+        let fee_rate_range = self.get_l1_fee_rate_range()?;
+
+        let mut l1_fee_rate = self
             .da_service
             .get_fee_rate()
             .await
             .map_err(|e| anyhow!(e))?;
 
+        l1_fee_rate = if l1_fee_rate < *fee_rate_range.start() {
```
Comment (kpp) [inline crates/sequencer/src/sequencer.rs:485]: Why not clamp? =)

File: crates/sequencer/src/sequencer.rs [hunk 2]
```diff
@@ -464,12 +474,22 @@ where
             last_finalized_height
         );
 
-        let l1_fee_rate = self
+        let fee_rate_range = self.get_l1_fee_rate_range()?;
+
+        let mut l1_fee_rate = self
```
Comment (kpp) [inline crates/sequencer/src/sequencer.rs:479]: rust allows "redefining" a variable. Remove this let and create a new l1_fee_rate below

File: crates/soft-confirmation-rule-enforcer/src/query.rs [hunk 3]
```diff
@@ -52,4 +54,30 @@ impl<C: Context, Da: DaSpec> SoftConfirmationRuleEnforcer<C, Da> {
...
+
+    /// function to get min and max for next L1 fee rate
+    pub fn get_next_min_max_l1_fee_rate(
+        &self,
+        working_set: &mut WorkingSet<C>,
+    ) -> RpcResult<RangeInclusive<u128>> {
+        let last_l1_fee_rate = self.last_l1_fee_rate.get(working_set).unwrap_or(0);
+
+        if last_l1_fee_rate == 0 {
+            // on the first soft confirmtaion, we don't have a last fee rate
```
Comment (kpp) [inline crates/soft-confirmation-rule-enforcer/src/query.rs:66]:
```suggestion
            // on the first soft confirmation, we don't have a last fee rate
```

File: crates/sequencer/src/sequencer.rs (hunk 1 above)
Comment (eyusufatik) [reply crates/sequencer/src/sequencer.rs:485]: bc I couldn't find the proper clamp 😓

File: crates/sequencer/src/sequencer.rs [hunk 4]
```diff
@@ -464,12 +474,22 @@ where
...
             .await
             .map_err(|e| anyhow!(e))?;
 
+        l1_fee_rate = if l1_fee_rate < *fee_rate_range.start() {
+            *fee_rate_range.start()
+        } else if l1_fee_rate > *fee_rate_range.end() {
+            *fee_rate_range.end()
+        } else {
+            l1_fee_rate
+        };
```
Comment (kpp) [inline crates/sequencer/src/sequencer.rs:491]:
```suggestion
        let fee_rate_range = self.get_l1_fee_rate_range()?;

        let l1_fee_rate = self
            .da_service
            .get_fee_rate()
            .await
            .map_err(|e| anyhow!(e))?;

        let l1_fee_rate = if l1_fee_rate < *fee_rate_range.start() {
            *fee_rate_range.start()
        } else if l1_fee_rate > *fee_rate_range.end() {
            *fee_rate_range.end()
        } else {
            l1_fee_rate
        };
```

**Review by kpp — APPROVED** [review]

**Pull request merged**